feat: add a more robust set of tests

hanoii opened the pull request:
Adding something more useful in terms of test coverage, for now this adds:

Different drupal version and a smaller set of tests are run by GHA strategy matrix

- a very bear dummy drupal module instead of relying on keycdn
- tests if the module has no composer ddev poseer should still work
- with both d10 and 11 of core supported versions tests that:
  -  it installs
  - ddev poser works
  - ddev symlink-project works
  - the core version install is the one specified
  - php tests comamnds are available
  - node commands are available
 
What I like to add in this initial approach is that eslint and stylelint actually does something, but that can be added later.

~The tests will fail until #81  and #80 are accepted.~ They are in, tests should now be green

hanoii commented:
tests are green

weitzman commented:
LGTM. 

I have done it in a while, but I want to make sure our tests can still pass locally. Could you please check that and document at the bottom of the README that command to run?

hanoii commented:
@weitzman I did run the tests manually several times but in testing this again I notices some potential issues by not using the new `TEST_DRUPAL_CORE` env var. 

I fixed this and now just running it locally works and uses the core default.

Added to the readme, latest version can be read on https://github.com/hanoii/ddev-drupal-contrib/blob/tests-rehaul/README.md#add-on-tests

hanoii commented:
Still doing some work

hanoii commented:
@weitzman it's ready for review, I decided to use git submodule for bash, see https://github.com/ddev/github-action-add-on-test/issues/36, but I don't any change on the ddev bats github action thanks to ` disable_checkout_action`.

My rationale for this is that it's a lot easier to install and to depend on things like bats-assert. Updated the README but the gist to run this locally is:

`git submodule init`
`./tests/bats/bin/bats tests`

Hope you like, it was quite a bit of work getting it right!

weitzman commented:
Any feedback, @rfay @tyler36?

tyler36 reviewed (commented):
Works well and as intended. 
I like the detailed test output.

I've had bad expereience with git submodules before so I tend to avoid them, however, it worked locally for me here. And more importantly didn't conflict with my local `bats` brew setup.

rfay commented on tests/full.bats:
```diff
@@ -0,0 +1,55 @@
+  fi
+  load '_common.bash'
+  _common_setup
+  ddev config --php-version=8.2
+  if [ "$TEST_DRUPAL_CORE" = "11" ]; then
+    ddev config --php-version=8.3 --corepack-enable
+  fi
```
I recommend just `ddev config --auto`, which does all these things.

```suggestion
  ddev config --auto
```

hanoii replied:
How does ddev config --auto does that? Basically I wanted to test d10 without corepack and d11 with, and also with both different php versions, not really needed, but figure it wasn't too bad

hanoii replied:
🤔 I think you implied that it does that be detecting d11, but d11 would not be installed yet on this case as it's taken care by `ddev poser` so I think this would not be the case.

hanoii replied:
I would leave it like this for now if that's ok, it is also more verbose.

rfay reviewed (changes requested):
Any time somebody goes to the crazy effort to expand and improve tests I have to be happy! 

I'm pretty sure we don't need specialty handling for d10/d11.

I wouldn't have put the bats extensions in git submodules because I've always hated them, but the README makes it sound easy to activate them.

hanoii commented on README.md:
```diff
@@ -104,6 +104,40 @@ ddev phpcbf -q
 
+## Add-on tests
+
+Tests are done with Bats. It is a simple testing framework that just uses Bash. 
```
I copied this over from https://github.com/ddev/ddev-addon-template, you might want to suggest edits there too 😄

tyler36 replied:
Ah, I thought it felt familiar. Thank you for the lead.

hanoii commented:
Most suggestion were addressed but one which I followed up. Thanks

rfay commented:
My suggestion of `ddev config --auto` was based on the assumption that the code is already laid down; it works well in that situation. DDEV detects docroot and subversion of `drupal` and does the right thing with corepack, etc. So I wasn't saying "don't test both d10 and d11", I was saying that if you like it, you can use DDEV's auto-config that detects and makes the right choices.

hanoii commented:
but the code is not already layed out in the context of that test, as we are testing adding the core later on as part of `ddev poser`

hanoii commented:
I was going to add some thoughts on the git submodules discussion, only because I feel it's interesting... but added it to https://github.com/ddev/ddev-addon-template/issues/62 instead

rfay commented:
If the code isn't there yet, then using `ddev config` to add hints of what's going on is a great technique, as in the quickstarts, which do the config before the composer create-project.  https://ddev.readthedocs.io/en/stable/users/quickstart/#drupal
